internal/topology/variables/clusterclass_variable_validation.go

```diff
@@ -215,9 +233,172 @@ func validateSchema(schema *apiextensions.JSONSchemaProps, fldPath *field.Path)
 		allErrs = append(allErrs, validateSchema(&p, fldPath.Child("properties").Key(propertyName))...)
 	}
 
```

Comment:
Hm I was looking at the upstream code and I'm really wondering why the recursion is so different here.

Upstream they basically
* call RootCELContext(schema) only once in validateCustomResourceDefinitionValidation (here this func would be validateRootSchema)
* then they:
  * Call ValidateCustomResourceDefinitionOpenAPISchema (here validateSchema) which calls itself resurcively for object, maps, arrays
  * The resursive calls of ValidateCustomResourceDefinitionOpenAPISchema are also getting the corresponding CEL context
  * Add the end of ValidateCustomResourceDefinitionOpenAPISchema they go through the validation of the CEL fields including the cost stuff
  * Add the end of validateCustomResourceDefinitionValidation they check the total cost
    * As far as I can tell our implementation adds to the TotalCost but we don't check it at the end


What we have here is basically that validateSchema resursively calls itself and on top of that validateCELExpressions calls itself recursively (with the addition of again calling validateSchema recursively for arrays).

At least calling validateSchema recursively within validateCELExpressions is definitely wrong. Also recursively calling validateCELExpressions for every schema seems wrong because we already recurse through the schema via validateSchema

Let's please align our recursion to the way it works in the upstream code?

Reply:
Updated the recursion, please take another look.